index.dev.js

```diff
@@ -25,56 +26,59 @@ function runInsect(fmt, line) {
   return res;
 }
 
-// Handle command line arguments
-if (process.argv.length >= 4) {
-  usage();
-} else if (process.argv.length == 3) {
-  var arg = process.argv[2];
-  if (arg === "-h" || arg === "--help") {
+// Top-level await is not supported in Node 12 and earlier.
+(async function() {
+  if (process.env.INSECT_NO_RC !== "true") {
+    function processRcLine(line) {
+      var res = runInsect(Insect.fmtPlain, line);
+      // We really only care when it breaks
+      if (res && res.msgType === "error") {
+        console.error(res.msg);
+        process.exit(1);
+      }
+    }
+
+    var [util, lineReader] = await Promise.all([import("util"), import("line-reader")]);
+    var eachLine = util.promisify(lineReader.eachLine);
+
+    try {
+      await eachLine(path.join(xdgBasedir.config, "insect/insectrc"), processRcLine);
+    } catch (err) {
+      if (err.code === "ENOENT") {
+        try {
+          var os = await import("os");
+          await eachLine(path.join(os.homedir(), ".insectrc"), processRcLine);
+        } catch (err) {
+          if (err.code !== "ENOENT") {
+            throw err;
+          }
+        }
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  // Handle command line arguments
+  var args = process.argv.slice(2);
+  if (args[0] === "-h" || args[0] === "--help") {
     usage();
-  } else {
+  } else if (args.length !== 0) {
     // Execute a single command
-    var res = runInsect(Insect.fmtPlain, arg);
+    var res = runInsect(Insect.fmtPlain, args.join(" "));
     if (res.msgType === "value" || res.msgType === "info") {
       console.log(res.msg);
     } else if (res.msgType === "error") {
       console.error(res.msg);
+      process.exit(1);
     }
     process.exit(0);
   }
-}
 
-if (process.env.INSECT_NO_RC !== "true") {
-  // Top-level await is not supported in Node 12 and earlier.
-  (async function() {
-    var [os, lineReader] = await Promise.all([import("os"), import("line-reader")]);
-
-    var rcFile = path.join(os.homedir(), ".insectrc");
-    lineReader.eachLine(rcFile, function (line) {
-      var res = runInsect(Insect.fmtPlain, line);
-      // We really only care when it breaks
-      if (res && res.msgType === "error") {
-        console.error(res.msg);
-        process.exit(1);
-      }
-    }, function (err) {
-      // If the file doesn't exist, that's fine
-      if (err && err.code !== "ENOENT") {
-        throw err;
-      } else {
-        startInsect();
-      }
-    });
-  })();
-} else {
-  startInsect();
-}
-
-async function startInsect() {
   var interactive = process.stdin.isTTY;
 
   if (interactive) {
-    var [fs, clipboardy, readline, xdgBasedir] = await Promise.all([import("fs"), import("clipboardy"), import("readline"), import("xdg-basedir")]);
+    var [fs, clipboardy, readline] = await Promise.all([import("fs"), import("clipboardy"), import("readline")]);
```

Comment:
I highly dislike the dynamic imports that we're doing here, but unfortunately they do help improve startup speeds by a good margin (I could get the exact numbers if you want). What do you think?

Reply:
If there are no downsides, I am in favor of the fast solution — even if the code could be much simpler :smile: